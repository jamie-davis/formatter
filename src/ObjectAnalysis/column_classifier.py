from collections import defaultdict
from enum import IntEnum, auto


class ColumnType(IntEnum):
    MIXED = auto()
    STRING = auto()
    NUMBER = auto()
    DATE = auto()
    UNKNOWN = auto()


def classify_columns(field_extract):
    return [ClassifiedColumn.classify(key, field_extract) for key in field_extract.keys]


def classify(value):
    types = {
        'int': ColumnType.NUMBER,
        'float': ColumnType.NUMBER,
        'str': ColumnType.STRING,
        'date': ColumnType.DATE,
        'datetime': ColumnType.DATE,
        'time': ColumnType.DATE,
        'complex': ColumnType.NUMBER
    }
    return types.get(type(value).__name__, ColumnType.UNKNOWN)


def get_types(values):
    column_types = defaultdict(int)
    for value in values:
        column_types[classify(value)] += 1
    return column_types


def calculate_width(values):
    return max([max([len(line) for line in str(value).splitlines()]) for value in values])


class ClassifiedColumn:
    def __init__(self, heading, left_align, required_width, column_type):
        self.heading = heading
        self.left_align = left_align
        self.required_width = required_width
        self.column_type = column_type

    @staticmethod
    def classify(key, field_extract):
        values = [row[key] for row in field_extract.rows if key in row.keys()]
        value_types = get_types(values)
        if len(value_types) > 1:
            column_type = ColumnType.MIXED
        else:
            column_type = next(iter(value_types))

        return ClassifiedColumn(key, column_type != ColumnType.NUMBER, max(len(key) if key is not None else 0, calculate_width(values)), column_type)
