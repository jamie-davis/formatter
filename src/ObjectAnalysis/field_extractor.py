def get_keys(value_pairs):
    return [pair for pair in value_pairs]


def primitive_dict(item):
    result = dict()
    result[None] = item
    return result


def get_values(item):
    if hasattr(item, '__dict__'):
        return vars(item)

    if isinstance(item, dict):
        return item

    return primitive_dict(item)


class FieldExtractor:
    """Extracts the fields present on a collection of objects."""

    @staticmethod
    def extract(items):
        """Analyse a list of objects and return a FieldExtractor instance containing the result"""
        type_name = type(items).__name__
        if type_name != "list" and type_name != "tuple":
            items = [items]

        all_values = [get_values(item) for item in items]

        extract = FieldExtractor()

        all_keys = [key for valuePairs in all_values for key in valuePairs]
        extract.keys = list(dict.fromkeys(all_keys))
        extract.rows = all_values
        return extract
