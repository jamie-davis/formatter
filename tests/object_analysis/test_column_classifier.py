import datetime

from approvaltests import verify

from ObjectAnalysis.column_classifier import classify, calculate_width


def test_column_type_classifiers():
    values = [1, 1.1, 'string', [],
              complex(5, 3),
              datetime.date(2020, 11, 18),
              datetime.datetime(2020, 11, 18, 11, 35, 44, 500000),
              datetime.time(11, 36, 1, 998999)]
    result = ''
    for value in values:
        result += str(value) + ' ' + type(value).__name__ + ': ' + str(classify(value)) + "\n"
    verify(result)


def test_calculate_width():
    values = [1, 1.1, [],
              complex(5, 3),
              datetime.date(2020, 11, 18),
              datetime.datetime(2020, 11, 18, 11, 35, 44, 500000),
              datetime.time(11, 36, 1, 998999)]
    expected = max([len(str(value)) for value in values])
    assert calculate_width(values) == expected


def test_calculate_width_finds_longest_line():
    values = ["1234567890",
              """1234567890
12345678901234567890
123456789012345678901234567890
12345678901234567890
1234567890"""]
    assert calculate_width(values) == 30
