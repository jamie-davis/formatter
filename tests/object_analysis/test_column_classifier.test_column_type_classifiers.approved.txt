1 int: ColumnType.NUMBER
1.1 float: ColumnType.NUMBER
string str: ColumnType.STRING
[] list: ColumnType.UNKNOWN
(5+3j) complex: ColumnType.NUMBER
2020-11-18 date: ColumnType.DATE
2020-11-18 11:35:44.500000 datetime: ColumnType.DATE
11:36:01.998999 time: ColumnType.DATE
